corehq/apps/userreports/indicators/__init__.py

```diff
@@ -130,13 +131,18 @@ def _make_column(self, product_code):
     @staticmethod
     def _get_values_by_product(ledger_section, case_id, product_codes, domain):
         """returns a defaultdict mapping product codes to their values"""
-        ret = defaultdict(lambda: 0)
         ledgers = LedgerAccessors(domain).get_ledger_values_for_case(case_id)
+        ledgers = [ledger for ledger in ledgers if ledger.section_id == ledger_section]
+        entry_ids = SQLProduct.objects.filter(
+            domain=domain, code__in=product_codes
+        ).values_list('product_id', flat=True)
+
+        ledger_values = defaultdict(lambda: 0)
         for ledger in ledgers:
-            if ledger.section_id == ledger_section and ledger.entry_id in product_codes:
-                ret[ledger.entry_id] = ledger.stock_on_hand
+            if ledger.entry_id in entry_ids:
```

Comment:
> It feels a little weird how ledgers is being filtered above in a list comprehension as well as here in a for loop. Would be nice to do those in the same place, or at least in the same format.

Fixed in new implementation

Regarding performance considerations: 

1) There are only two data sources that currently use this on HQ, so this isn't hit very often. 
2) For ICDS, they use ledgers as a part of the due list where every immunization is a product. There's less than 50 and we'll need to track all of them, so it doesn't matter that it's being filtered in memory since they'll need to be in memory anyways.

I started to fix this by adding filters for entity ID and section ID to `get_ledger_values_for_case`, but stopped because, to have this be one query for couch domains, I'd need to insert knowledge about products in `LedgerAccessors` which didn't seem ideal. For SQL, we can't do it in one query on the partitioned tables because ledger values are separated from products and for ICDS we need all the ledgers anyways (point 2), so it didn't seem worthwhile to make the effort to make a SQL function.

Reply:
Cool, that all sounds reasonable.